src/xcb_ffi/raw_ffi.rs

```diff
@@ -129,57 +132,182 @@ pub(crate) mod send_request_flags {
     pub(crate) const REPLY_FDS: c_int = 8;
 }
 
+#[cfg(all(not(test), feature = "dl-libxcb"))]
+struct LibxcbLibrary {
+    // Needed to keep the library loaded
+    _library: libloading::Library,
+    funcs: LibxcbFuncs,
+}
+
+#[cfg(all(not(test), feature = "dl-libxcb"))]
+impl LibxcbLibrary {
+    fn open_lib() -> Result<libloading::Library, LibxcbLoadError> {
+        // TODO: Names for non-unix platforms
+        #[cfg(unix)]
+        const LIB_NAMES: &[&str] = &["libxcb.so.1", "libxcb.so"];
+        #[cfg(not(unix))]
+        compile_error!("dl-libxcb feature is not supported on non-unix");
+
+        let mut errors = Vec::new();
+        for lib_name in LIB_NAMES.iter() {
+            match libloading::Library::new(lib_name) {
+                Ok(library) => return Ok(library),
+                Err(e) => errors.push((lib_name.into(), e.to_string())),
+            }
+        }
+        Err(LibxcbLoadError::OpenLibError(errors))
+    }
+
+    /// # Safety
+    ///
+    /// The functions pointers in `funcs` do not have lifetime,
+    /// but they must not outlive the returned result.
+    #[cold]
+    #[inline(never)]
+    unsafe fn load() -> Result<Self, LibxcbLoadError> {
+        let library = Self::open_lib()?;
+        let funcs = LibxcbFuncs::new(&library)
+            .map_err(|(symbol, e)| LibxcbLoadError::GetSymbolError(symbol.into(), e.to_string()))?;
+        Ok(Self {
+            _library: library,
+            funcs,
+        })
+    }
+}
+
+#[cfg(all(not(test), feature = "dl-libxcb"))]
```

Comment:
How about moving things into an extra module so that only one instance of this repeated `#[cfg]` is necessary?

Reply:
Not a bad idea.